// create-readme/lib/components/badges.js
// Generated by CoffeeScript 1.10.0
(function() {
  var BadgeParser, Promise, badgeData, logger, mustache;

  logger = require('../logger');

  Promise = require('bluebird');

  mustache = require('mustache');

  badgeData = require('../../data/badges.json');

  module.exports = BadgeParser = (function() {
    BadgeParser.name = "BadgeParser";

    function BadgeParser(options) {
      var base;
      this.options = options;
      if (this.options == null) {
        this.options = {};
      }
      if ((base = this.options).badges == null) {
        base.badges = ['npm-version', 'travis', 'coveralls', 'dependencies', 'devDependencies', 'gitter'];
      }
    }

    BadgeParser.prototype.run = function(pkg) {
      logger.info("Creating badges");
      if (pkg.git == null) {
        logger.debug(" Not adding badges due to missing git repo info");
        return Promise.resolve([]);
      }
      logger.debug(" Adding badges " + this.options.badges);
      return Promise.resolve(this.options.badges.map(function(badgeName) {
        return {
          name: mustache.render(badgeData[badgeName].name, pkg),
          img: mustache.render(badgeData[badgeName].img, pkg),
          url: mustache.render(badgeData[badgeName].url, pkg)
        };
      }));
    };

    return BadgeParser;

  })();

}).call(this);
