// create-readme/lib/components/install.js
// Generated by CoffeeScript 1.10.0
(function() {
  var InstallParser, Promise, logger;

  logger = require('../logger');

  Promise = require('bluebird');

  module.exports = InstallParser = (function() {
    InstallParser.name = "InstallParser";

    function InstallParser(options) {
      var base, base1;
      this.options = options;
      if (this.options == null) {
        this.options = {};
      }
      if ((base = this.options).modules == null) {
        base.modules = ['CommonJS'];
      }
      if ((base1 = this.options).npmcdn == null) {
        base1.npmcdn = false;
      }
    }

    InstallParser.prototype.run = function(pkg) {
      logger.info("Creating installation info");
      logger.debug(" Adding modules " + this.options.modules);
      if (this.options.npmcdn) {
        logger.debug(" Adding npmcdn");
      }
      return Promise.resolve({
        modules: {
          names: this.options.modules.map(function(module) {
            return {
              name: module
            };
          })
        },
        npmcdn: this.options.npmcdn
      });
    };

    return InstallParser;

  })();

}).call(this);
