// create-readme/lib/components/usage.js
// Generated by CoffeeScript 1.10.0
(function() {
  var Promise, UsageParser, fs, languages, logger, path;

  logger = require('../logger');

  Promise = require('bluebird');

  fs = Promise.promisifyAll(require('fs'));

  path = require('path');

  languages = require('../../data/languages.json');

  module.exports = UsageParser = (function() {
    UsageParser.name = "UsageParser";

    function UsageParser(options) {
      var base, base1, base2, base3;
      this.options = options;
      if (this.options == null) {
        this.options = {};
      }
      if ((base = this.options).encoding == null) {
        base.encoding = 'utf-8';
      }
      if ((base1 = this.options).replaceReferences == null) {
        base1.replaceReferences = true;
      }
      if ((base2 = this.options).addUsage == null) {
        base2.addUsage = '';
      }
      if ((base3 = this.options).exampleDir == null) {
        base3.exampleDir = "examples";
      }
    }

    UsageParser.prototype.run = function(pkg) {
      var exampleDir, ref, ref1;
      logger.info("Creating usage examples");
      exampleDir = (ref = (ref1 = pkg.directories) != null ? ref1.example : void 0) != null ? ref : this.options.exampleDir;
      exampleDir = path.join(process.cwd(), exampleDir);
      logger.debug(" Looking for examples in dir " + exampleDir);
      return fs.readdirAsync(exampleDir)["catch"]({
        code: 'ENOENT'
      }, function() {
        return [];
      }).filter(function(file) {
        return Object.keys(languages).indexOf(path.extname(file)) > -1;
      }).map((function(_this) {
        return function(file) {
          var content;
          logger.debug(" Parsing file" + file);
          content = fs.readFileAsync(path.join(exampleDir, file), {
            encoding: _this.options.encoding
          }).then(function(content) {
            var regex;
            if (!_this.options.replaceReferences) {
              return content;
            }
            logger.debug(" Replacing occurances of '../' and \"../\" with " + pkg.name);
            regex = /(?:'\.\.\/')|(?:"\.\.\/")/g;
            return content.replace(regex, "'" + pkg.name + "'");
          });
          return Promise.resolve({
            lang: languages[path.extname(file)],
            content: content
          }).props();
        };
      })(this)).then((function(_this) {
        return function(examples) {
          var ref2;
          if ((examples.length === (ref2 = _this.options.addUsage.length) && ref2 === 0)) {
            logger.debug(" No examples and no addUsage");
            return null;
          } else {
            return {
              examples: examples,
              description: _this.options.addUsage
            };
          }
        };
      })(this));
    };

    return UsageParser;

  })();

}).call(this);
